onnx/defs/gen_doc.py

```diff
@@ -79,6 +79,25 @@ def display_version_link(name, version):  # type: (Text, int) -> Text
     return '<a href="{}#{}">{}</a>'.format(changelog_md, name_with_ver, name_with_ver)
 
 
+def generate_formal_parameter_tags(formal_parameter):  # type: (OpSchema.FormalParameter) -> Text
+    tags = []  # type: List[Text]
+    if OpSchema.FormalParameterOption.Optional == formal_parameter.option:
+        tags = ["optional"]
+    elif OpSchema.FormalParameterOption.Variadic == formal_parameter.option:
+        if formal_parameter.isHomogeneous:
+            tags = ["variadic"]
+        else:
+            tags = ["variadic", "heterogeneous"]
+    differentiable = OpSchema.DifferentiationCategory.Differentiable  # type: OpSchema.DifferentiationCategory
+    non_differentiable = OpSchema.DifferentiationCategory.NonDifferentiable  # type: OpSchema.DifferentiationCategory
+    if differentiable == formal_parameter.differentiationCategory:
+        tags.append('differentiable')
+    elif non_differentiable == formal_parameter.differentiationCategory:
+        tags.append('non-differentiable')
```

Comment:
how about "unknown" case?

Reply:
I'm wondering whether a separate "training" op doc should be generated, that will make the doc understanding easier, given ONNX may support more scenarios, other than DNN. thoughts?

Reply:
Unknown would be an empty string. Would it be better to use `differentiability` unknown`? 

For the second comment, I don't feel splitting this tag from `Operator.md` is a good idea. To understand why an input/output is differentiable and how to compute its gradient, the reader must read that operator's entire document.

Reply:
I don't believe that training algorithms used for deep learning training can be applied to many traditional ML models. Hence, we probably don't need to add differentiability tags to ops in ONNX-ML.

Reply:
@wschin I'm ok for "differentialbility - unknown" or "undefined" to make the 3 cases very clear - "differentiable, not differentiable, undefined", if we don't want to put the training ops into a separate doc.

Reply:
or maybe one line of "statement" - inputs without specifying "differentiability" have that as "undefined" at the beginning of operators.md?

Reply:
@linkerzhang, training ops may not be put in another MD because --- this newly added attribute will be added to ALL existing operators. 

Because each operator has their own differentiability, I am not sure how to create an one-line statement in the beginning of `operator.md`. Do you mean create one line for each operator?

Reply:
yes, I mean one line in the beginning of operator.md to save clarifying it in each operator :).

Reply:
@linkerzhang, I think I get your point. Is https://github.com/onnx/onnx/pull/2723/commits/1f1fc67cd2eefc3144841859255a81280f66dff7 we want?